File: src/programmers/helpers/disable_function_programmer_declare.ts
import ts from "typescript";

import { FunctionProgrammer } from "./FunctionProgrammer";

export const disable_function_programmer_declare = (
  functor: FunctionProgrammer,
): FunctionProgrammer => disable(functor) as FunctionProgrammer;

const disable = (
  functor: FunctionProgrammer,
): MethodOnly<FunctionProgrammer> => ({
  method: functor.method,
  useLocal: (name: string): string => functor.useLocal(name),
  hasLocal: (name: string): boolean => functor.hasLocal(name),
  declare: (): ts.Statement[] => [],
  declareUnions: (): ts.Statement[] => [],
  increment: (): number => functor.increment(),
  emplaceUnion: (
    prefix: string,
    name: string,
    factory: () => ts.ArrowFunction,
  ): string => functor.emplaceUnion(prefix, name, factory),
  emplaceVariable: (key, value) => functor.emplaceVariable(key, value),
});

type MethodOnly<T> = {
  [P in keyof T]: T[P] extends Function
    ? T[P]
    : P extends "method"
      ? T[P]
      : never;
};

File: packages/core/src/programmers/helpers/disable_function_programmer_declare.go
package helpers

type DisabledFunctionProgrammer struct {
	Origin *FunctionProgrammer
}

func DisableFunctionProgrammerDeclare(programmer *FunctionProgrammer) DisabledFunctionProgrammer {
	return DisabledFunctionProgrammer{Origin: programmer}
}

func (p DisabledFunctionProgrammer) UseLocal(name string) string {
	return p.Origin.UseLocal(name)
}

func (p DisabledFunctionProgrammer) HasLocal(name string) bool {
	return p.Origin.HasLocal(name)
}

func (p DisabledFunctionProgrammer) Declare(name string, body ConciseBody) {}

func (p DisabledFunctionProgrammer) DeclareUnions() []Statement { return nil }

func (p DisabledFunctionProgrammer) Increment() int {
	return p.Origin.Increment()
}

func (p DisabledFunctionProgrammer) EmplaceUnion(prefix string, body []Statement) string {
	return p.Origin.EmplaceUnion(prefix, body)
}

func (p DisabledFunctionProgrammer) EmplaceVariable(prefix string, value Expression) string {
	return p.Origin.EmplaceVariable(prefix, value)
}
